src/utils/streams/promise_from_streams.js

```diff
@@ -33,54 +33,30 @@
  *  @param  {Array<Stream>} streams
  *  @return {Promise<any>}
  */
-export async function createPromiseFromStreams(streams) {
-  const last = streams[streams.length - 1];
 
-  // reject if any of the streams emits an error
-  const anyStreamFailure = new Promise((resolve, reject) => {
-    streams.forEach((stream, i) => {
-      if (i > 0) streams[i - 1].pipe(stream);
-      stream.on('error', reject);
-      return stream;
-    });
-  });
-
-  // resolve when the last stream has finished writing, or
-  // immediately if the last stream is not writable
-  const lastFinishedWriting = new Promise(resolve => {
-    if (typeof last.write !== 'function') {
-      resolve();
-      return;
-    }
-
-    last.on('finish', resolve);
-  });
+import { pipeline, Writable } from 'stream';
 
-  // resolve with the final value provided by the last stream
-  // after the last stream has provided it, or immediately if the
-  // stream is not readable
-  const lastFinishedReading = new Promise(resolve => {
-    if (typeof last.read !== 'function') {
-      resolve();
-      return;
-    }
-
-    let finalChunk;
-    last.on('data', (chunk) => {
-      finalChunk = chunk;
-    });
-    last.on('end', () => {
+export async function createPromiseFromStreams(streams) {
+  let finalChunk;
+  const last = streams[streams.length - 1];
+  if (typeof last.read === 'function') {
+    // We are pushing a writable stream to capture the last chunk
+    // but also to ensure we have the minimum 2 streams requirement
+    // for stream.pipeline.
+    streams.push(new Writable({
+      // Use object mode even when "last" stream isn't. This allows to
+      // capture the last chunk as-is.
+      objectMode: true,
+      write(chunk, enc, done) {
+        finalChunk = chunk;
+        done();
+      }
+    }));
+  }
```

Comment:
Do you think we should add a check for `typeof last.read !== 'function'` && streams.length === 1? We might be able to help out with a slightly more informative error message than whatever stream.pipeline produces by default.

Reply:
I agree, it did take me a bit to understand the default error. I made the change here https://github.com/elastic/kibana/pull/27246/commits/cd3392fedfd7edb751e79a1c4908fa129e5aa34b, let me know if the message makes sense to you.

Reply:
Error message looks good, but go ahead and just throw the error, the async function will automatically reject the promise